src/json.ts

```diff
@@ -4,9 +4,110 @@ export type JSONValue =
   | boolean
   | number
   | Array<JSONValue>
-  // undefined are filtered out in JSON.stringify so OK to allow.
-  | {[key: string]: JSONValue | undefined};
+  | JSONObject;
+
+type JSONObject = {[key: string]: JSONValue};
 
 export type ToJSON = {
-  toJSON: () => JSONValue;
+  toJSON: (key?: string) => JSONValue;
 };
+
+/**
+ * Like [[JSONValue]] but allows [[ToJSON]] at every recursive value. This
+ * matches the semantics of `JSON.stringify`.
+ */
+export type ToJSONValue =
+  | ToJSON
+  | null
+  | string
+  | boolean
+  | number
+  | Array<ToJSONValue>
+  | ToJSONObject;
+
+type ToJSONObject = {[key: string]: ToJSONValue};
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+function hasToJSON(v: any): v is ToJSON {
+  return typeof v?.toJSON === 'function';
+}
+
+/**
+ * Does a deep equality check on a JSON value with same semantics as
+ * JSON.stringify (except for the ordering of the keys in an Object).
+ */
+export function deepEqual(
+  a: ToJSONValue | undefined,
+  b: ToJSONValue | undefined,
+): boolean {
+  return deepEqualWithKey(a, b, '');
+}
+
+function deepEqualWithKey(
+  a: ToJSONValue | undefined,
+  b: ToJSONValue | undefined,
+  key: string,
+): boolean {
+  if (a === b) {
+    return true;
+  }
+
+  if (hasToJSON(a)) {
+    return deepEqualWithKey(b, a.toJSON(key), key);
```

Comment:
I don't like the idea of accepting `ToJSON` here because it means we end up doing work that is wasted to check for equality. Also there's no guarantee that the equality check works correctly because `toJSON()` could accidentally return something logically different than value it's a method of.

Reply:
I don't think that is a concern. toJSON is used in Date to allow some flexibility. But I'm in favor of simplifying this. I'm removing ToJSON here.